src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Controls/TitleBar.cs

```diff
@@ -0,0 +1,659 @@
+    }
+
+    /// <summary>
+    /// Event triggered after clicking minimize button.
+    /// </summary>
+    public event TypedEventHandler<TitleBar, RoutedEventArgs> MinimizeClicked
+    {
+        add => AddHandler(MinimizeClickedEvent, value);
+        remove => RemoveHandler(MinimizeClickedEvent, value);
+    }
+
+    /// <summary>
+    /// Event triggered after clicking help button
+    /// </summary>
+    public event TypedEventHandler<TitleBar, RoutedEventArgs> HelpClicked
+    {
+        add => AddHandler(HelpClickedEvent, value);
+        remove => RemoveHandler(HelpClickedEvent, value);
+    }
+
+    /// <summary>
+    /// Command triggered after clicking the titlebar button.
+    /// </summary>
+    public IRelayCommand TemplateButtonCommand => (IRelayCommand)GetValue(TemplateButtonCommandProperty);
+
+    /// <summary>
+    /// Lets you override the behavior of the Maximize/Restore button with an <see cref="Action"/>.
+    /// </summary>
+    public Action<TitleBar, System.Windows.Window> MaximizeActionOverride { get; set; }
+
+    /// <summary>
+    /// Lets you override the behavior of the Minimize button with an <see cref="Action"/>.
+    /// </summary>
+    public Action<TitleBar, System.Windows.Window> MinimizeActionOverride { get; set; }
+
+    #endregion
+
+    private System.Windows.Window _currentWindow = null!;
+    private System.Windows.Controls.Grid _mainGrid = null!;
+    private System.Windows.Controls.ContentPresenter _icon = null!;
+    private readonly TitleBarButton[] _buttons = new TitleBarButton[4];
+
+    /// <summary>
+    /// Creates a new instance of the class and sets the default <see cref="FrameworkElement.Loaded"/> event.
+    /// </summary>
+    public TitleBar()
+    {
+        SetValue(TemplateButtonCommandProperty, new RelayCommand<TitleBarButtonType>(OnTemplateButtonClick));
+
+        Loaded += OnLoaded;
+        Unloaded += OnUnloaded;
+    }
+
+    /// <inheritdoc />
+    protected override void OnInitialized(EventArgs e)
+    {
+        base.OnInitialized(e);
+
+        ApplicationTheme = Appearance.ApplicationThemeManager.GetAppTheme();
+        Appearance.ApplicationThemeManager.Changed += OnThemeChanged;
+    }
+
+    protected virtual void OnLoaded(object sender, RoutedEventArgs e)
+    {
+        // if (DesignerHelper.IsInDesignMode)
+        // {
+        //     return;
+        // }
+
+        _currentWindow =
+            System.Windows.Window.GetWindow(this) ?? throw new ArgumentNullException("Window is null");
+        _currentWindow.StateChanged += OnParentWindowStateChanged;
+
+        var handle = new WindowInteropHelper(_currentWindow).EnsureHandle();
+        var windowSource =
+            HwndSource.FromHwnd(handle) ?? throw new ArgumentNullException("Window source is null");
+        windowSource.AddHook(HwndSourceHook);
+    }
+
+    private void OnUnloaded(object sender, RoutedEventArgs e)
+    {
+        Loaded -= OnLoaded;
```

Comment:
I don't think `-=` is an appropriate action here, as it implies that once it's Unloaded, it will no longer listen for window changes when Loaded next time.